In [1]:
import pandas as pd

In [2]:
data=pd.read_csv("Salary.csv")

C:\Users\versha pande\AppData\Local\Temp\ipykernel_47296\44852249.py:1: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  data=pd.read_csv("Salary.csv")


In [3]:
data

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle       BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY     167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)     155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)     212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC       77916.0   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)      134401.6   
...                                                ...           ...   
148649                                       Custodian          0.00   
148650                                    Not provided  Not Provided   
148651                                    Not provided  Not Provided   
148652                                    Not provided  Not Provided   
148653                      Counselor, Log Cabin Ranch          0.00   

         OvertimePay      OtherPay      Benefits   TotalPay  TotalPayBenefits  \
0                0.0     400184.25           NaN  567595.43         567595.43   
1          245131.88     137811.38           NaN  538909.28         538909.28   
2          106088.18       16452.6           NaN  335279.91         335279.91   
3           56120.71      198306.9           NaN  332343.61         332343.61   
4             9737.0     182234.59           NaN  326373.19         326373.19   
...              ...           ...           ...        ...               ...   
148649          0.00          0.00          0.00       0.00              0.00   
148650  Not Provided  Not Provided  Not Provided       0.00              0.00   
148651  Not Provided  Not Provided  Not Provided       0.00              0.00   
148652  Not Provided  Not Provided  Not Provided       0.00              0.00   
148653          0.00       -618.13          0.00    -618.13           -618.13   

        Year  Notes         Agency Status  
0       2011    NaN  San Francisco    NaN  
1       2011    NaN  San Francisco    NaN  
2       2011    NaN  San Francisco    NaN  
3       2011    NaN  San Francisco    NaN  
4       2011    NaN  San Francisco    NaN  
...      ...    ...            ...    ...  
148649  2014    NaN  San Francisco     PT  
148650  2014    NaN  San Francisco    NaN  
148651  2014    NaN  San Francisco    NaN  
148652  2014    NaN  San Francisco    NaN  
148653  2014    NaN  San Francisco     PT  

[148654 rows x 13 columns]

## Display top 10 rows of the dataset:

In [4]:
data.head(10)

Id        EmployeeName                                        JobTitle  \
0   1      NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2        GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3      ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4   CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5     PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
5   6      DAVID SULLIVAN                       ASSISTANT DEPUTY CHIEF II   
6   7           ALSON LEE              BATTALION CHIEF, (FIRE DEPARTMENT)   
7   8       DAVID KUSHNER                  DEPUTY DIRECTOR OF INVESTMENTS   
8   9      MICHAEL MORRIS              BATTALION CHIEF, (FIRE DEPARTMENT)   
9  10  JOANNE HAYES-WHITE          CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   
5   118602.0      8601.0  189082.74      NaN  316285.74         316285.74   
6   92492.01     89062.9  134426.14      NaN  315981.05         315981.05   
7  256576.96         0.0    51322.5      NaN  307899.46         307899.46   
8  176932.64    86362.68   40132.23      NaN  303427.55         303427.55   
9   285262.0         0.0   17115.73      NaN  302377.73         302377.73   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN  
5  2011    NaN  San Francisco    NaN  
6  2011    NaN  San Francisco    NaN  
7  2011    NaN  San Francisco    NaN  
8  2011    NaN  San Francisco    NaN  
9  2011    NaN  San Francisco    NaN

## Display last 10 rows of the dataset:

In [5]:
data.tail(10)

Id      EmployeeName                      JobTitle       BasePay  \
148644  148645      Randy D Winn  Stationary Eng, Sewage Plant          0.00   
148645  148646  Carolyn A Wilson     Human Services Technician          0.00   
148646  148647      Not provided                  Not provided  Not Provided   
148647  148648    Joann Anderson   Communications Dispatcher 2          0.00   
148648  148649       Leon Walker                     Custodian          0.00   
148649  148650     Roy I Tillery                     Custodian          0.00   
148650  148651      Not provided                  Not provided  Not Provided   
148651  148652      Not provided                  Not provided  Not Provided   
148652  148653      Not provided                  Not provided  Not Provided   
148653  148654         Joe Lopez    Counselor, Log Cabin Ranch          0.00   

         OvertimePay      OtherPay      Benefits  TotalPay  TotalPayBenefits  \
148644          0.00          0.00          0.00      0.00              0.00   
148645          0.00          0.00          0.00      0.00              0.00   
148646  Not Provided  Not Provided  Not Provided      0.00              0.00   
148647          0.00          0.00          0.00      0.00              0.00   
148648          0.00          0.00          0.00      0.00              0.00   
148649          0.00          0.00          0.00      0.00              0.00   
148650  Not Provided  Not Provided  Not Provided      0.00              0.00   
148651  Not Provided  Not Provided  Not Provided      0.00              0.00   
148652  Not Provided  Not Provided  Not Provided      0.00              0.00   
148653          0.00       -618.13          0.00   -618.13           -618.13   

        Year  Notes         Agency Status  
148644  2014    NaN  San Francisco     PT  
148645  2014    NaN  San Francisco     PT  
148646  2014    NaN  San Francisco    NaN  
148647  2014    NaN  San Francisco     PT  
148648  2014    NaN  San Francisco     PT  
148649  2014    NaN  San Francisco     PT  
148650  2014    NaN  San Francisco    NaN  
148651  2014    NaN  San Francisco    NaN  
148652  2014    NaN  San Francisco    NaN  
148653  2014    NaN  San Francisco     PT

## Find shape of the dataset:

In [6]:
data.shape

(148654, 13)

In [7]:
# Now here we see output as a tuple:
print("Numnber of rows:",data.shape[0])
print("Numnber of columns:",data.shape[1])

Numnber of rows: 148654
Numnber of columns: 13


In [8]:
data.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [9]:
data.index

RangeIndex(start=0, stop=148654, step=1)

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


## Checking the null value in the dataset:

In [11]:
data.isnull()

Id  EmployeeName  JobTitle  BasePay  OvertimePay  OtherPay  \
0       False         False     False    False        False     False   
1       False         False     False    False        False     False   
2       False         False     False    False        False     False   
3       False         False     False    False        False     False   
4       False         False     False    False        False     False   
...       ...           ...       ...      ...          ...       ...   
148649  False         False     False    False        False     False   
148650  False         False     False    False        False     False   
148651  False         False     False    False        False     False   
148652  False         False     False    False        False     False   
148653  False         False     False    False        False     False   

        Benefits  TotalPay  TotalPayBenefits   Year  Notes  Agency  Status  
0           True     False             False  False   True   False    True  
1           True     False             False  False   True   False    True  
2           True     False             False  False   True   False    True  
3           True     False             False  False   True   False    True  
4           True     False             False  False   True   False    True  
...          ...       ...               ...    ...    ...     ...     ...  
148649     False     False             False  False   True   False   False  
148650     False     False             False  False   True   False    True  
148651     False     False             False  False   True   False    True  
148652     False     False             False  False   True   False    True  
148653     False     False             False  False   True   False   False  

[148654 rows x 13 columns]

In [12]:
len(data.isnull())

148654

In [13]:
data.isnull().count

<bound method DataFrame.count of            Id  EmployeeName  JobTitle  BasePay  OvertimePay  OtherPay  \
0       False         False     False    False        False     False   
1       False         False     False    False        False     False   
2       False         False     False    False        False     False   
3       False         False     False    False        False     False   
4       False         False     False    False        False     False   
...       ...           ...       ...      ...          ...       ...   
148649  False         False     False    False        False     False   
148650  False         False     False    False        False     False   
148651  False         False     False    False        False     False   
148652  False         False     False    False        False     False   
148653  False         False     False    False        False     False   

        Benefits  TotalPay  TotalPayBenefits   Year  Notes  Agency  Status  
0           T

In [14]:
# Best way:
data.isnull().sum()
# Base pay have 605 null values.

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

## Drop unwanted columns in the dataset:

In [15]:
data.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [16]:
# We can drop unwanted columns for better analysis.
data=data.drop(['Id','Notes', 'Agency','Status'],axis=1)

In [17]:
data.head(1)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   

  OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  Year  
0         0.0  400184.25      NaN  567595.43         567595.43  2011

## Overall statistics of the dataset:

In [18]:
data.describe(include='all')
# By default all numerical by using include all we are getting categorical too.

EmployeeName          JobTitle   BasePay  OvertimePay  OtherPay  \
count        148654            148654  148049.0     148654.0  148654.0   
unique       110811              2159  109900.0      66555.0   84968.0   
top       Kevin Lee  Transit Operator       0.0          0.0       0.0   
freq             13              7036     875.0      66103.0   35218.0   
mean            NaN               NaN       NaN          NaN       NaN   
std             NaN               NaN       NaN          NaN       NaN   
min             NaN               NaN       NaN          NaN       NaN   
25%             NaN               NaN       NaN          NaN       NaN   
50%             NaN               NaN       NaN          NaN       NaN   
75%             NaN               NaN       NaN          NaN       NaN   
max             NaN               NaN       NaN          NaN       NaN   

        Benefits       TotalPay  TotalPayBenefits           Year  
count   112495.0  148654.000000     148654.000000  148654.000000  
unique   99635.0            NaN               NaN            NaN  
top          0.0            NaN               NaN            NaN  
freq      1053.0            NaN               NaN            NaN  
mean         NaN   74768.321972      93692.554811    2012.522643  
std          NaN   50517.005274      62793.533483       1.117538  
min          NaN    -618.130000       -618.130000    2011.000000  
25%          NaN   36168.995000      44065.650000    2012.000000  
50%          NaN   71426.610000      92404.090000    2013.000000  
75%          NaN  105839.135000     132876.450000    2014.000000  
max          NaN  567595.430000     567595.430000    2014.000000

## Find occurrance of the employee names top 5:

In [19]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [20]:
data['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650         Not provided
148651         Not provided
148652         Not provided
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [21]:
data['EmployeeName'].value_counts()

Kevin Lee                   13
Richard Lee                 11
Steven Lee                  11
William Wong                11
Stanley Lee                  9
                            ..
Jeffrey Skover               1
Ken Dever                    1
Kevin Whitfield              1
Granville McCollough III     1
Joe Lopez                    1
Name: EmployeeName, Length: 110811, dtype: int64

In [22]:
data['EmployeeName'].value_counts().head(5)

Kevin Lee       13
Richard Lee     11
Steven Lee      11
William Wong    11
Stanley Lee      9
Name: EmployeeName, dtype: int64

## Find the number of unique job titles:

In [23]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [24]:
data['JobTitle']

0         GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
1                        CAPTAIN III (POLICE DEPARTMENT)
2                        CAPTAIN III (POLICE DEPARTMENT)
3                   WIRE ROPE CABLE MAINTENANCE MECHANIC
4           DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)
                               ...                      
148649                                         Custodian
148650                                      Not provided
148651                                      Not provided
148652                                      Not provided
148653                        Counselor, Log Cabin Ranch
Name: JobTitle, Length: 148654, dtype: object

In [25]:
data['JobTitle'].unique()

array(['GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY',
       'CAPTAIN III (POLICE DEPARTMENT)',
       'WIRE ROPE CABLE MAINTENANCE MECHANIC', ..., 'Conversion',
       'Cashier 3', 'Not provided'], dtype=object)

In [26]:
len(data['JobTitle'].unique())

2159

In [27]:
# OR
data['JobTitle'].nunique()

2159

## Total number of job titles which contain captain:

In [28]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [29]:
data['JobTitle']

0         GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
1                        CAPTAIN III (POLICE DEPARTMENT)
2                        CAPTAIN III (POLICE DEPARTMENT)
3                   WIRE ROPE CABLE MAINTENANCE MECHANIC
4           DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)
                               ...                      
148649                                         Custodian
148650                                      Not provided
148651                                      Not provided
148652                                      Not provided
148653                        Counselor, Log Cabin Ranch
Name: JobTitle, Length: 148654, dtype: object

In [30]:
data['JobTitle'].str.contains('CAPTAIN')

0         False
1          True
2          True
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [31]:
data[data['JobTitle'].str.contains('CAPTAIN')]

EmployeeName                               JobTitle    BasePay  \
1          GARY JIMENEZ        CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2        ALBERT PARDINI        CAPTAIN III (POLICE DEPARTMENT)  212739.13   
11     PATRICIA JACKSON        CAPTAIN III (POLICE DEPARTMENT)    99722.0   
17       SEBASTIAN WONG  CAPTAIN, EMERGENCYCY MEDICAL SERVICES  140546.87   
22        GEORGE GARCIA              CAPTAIN, FIRE SUPPRESSION  140546.88   
...                 ...                                    ...        ...   
8684      JEANNE SEYLER              CAPTAIN, FIRE SUPPRESSION   95055.34   
10485        JANE SMITH  CAPTAIN, EMERGENCYCY MEDICAL SERVICES    74592.0   
11198     KATHRYN BROWN        CAPTAIN III (POLICE DEPARTMENT)    10684.5   
31297   MARCO CARNIGLIA  CAPTAIN, EMERGENCYCY MEDICAL SERVICES    9839.72   
34124     JOHN FORBES-3              CAPTAIN, FIRE SUPPRESSION        0.0   

      OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  Year  
1       245131.88  137811.38      NaN  538909.28         538909.28  2011  
2       106088.18    16452.6      NaN  335279.91         335279.91  2011  
11       87082.62   110804.3      NaN  297608.92         297608.92  2011  
17      119397.26   18625.08      NaN  278569.21         278569.21  2011  
22       93200.58   39955.25      NaN  273702.71         273702.71  2011  
...           ...        ...      ...        ...               ...   ...  
8684          0.0    9197.14      NaN  104252.48         104252.48  2011  
10485     1538.59   18804.68      NaN   94935.27          94935.27  2011  
11198         0.0   81244.87      NaN   91929.37          91929.37  2011  
31297         0.0    1203.77      NaN   11043.49          11043.49  2011  
34124      982.06    2277.34      NaN    3259.40           3259.40  2011  

[141 rows x 9 columns]

In [32]:
len(data['JobTitle'].str.contains('CAPTAIN'))
# There is a difference because this is a boolean serien for all the data in a column.

148654

In [33]:
data['JobTitle'].str.contains('CAPTAIN').count()

148654

In [34]:
data[data['JobTitle'].str.contains('CAPTAIN')].count()
# Here we have the rows which are having value as True.

EmployeeName        141
JobTitle            141
BasePay             141
OvertimePay         141
OtherPay            141
Benefits              0
TotalPay            141
TotalPayBenefits    141
Year                141
dtype: int64

In [35]:
len(data[data['JobTitle'].str.contains('CAPTAIN')])

141

### Above was case sensitive : to make it insensitive we need case=False

In [36]:
data['JobTitle'].str.contains('CAPTAIN',case=False)

0         False
1          True
2          True
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [37]:
data[data['JobTitle'].str.contains('CAPTAIN',case=False)]

EmployeeName                               JobTitle    BasePay  \
1             GARY JIMENEZ        CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2           ALBERT PARDINI        CAPTAIN III (POLICE DEPARTMENT)  212739.13   
11        PATRICIA JACKSON        CAPTAIN III (POLICE DEPARTMENT)    99722.0   
17          SEBASTIAN WONG  CAPTAIN, EMERGENCYCY MEDICAL SERVICES  140546.87   
22           GEORGE GARCIA              CAPTAIN, FIRE SUPPRESSION  140546.88   
...                    ...                                    ...        ...   
116604   Gregory L Stewart              Captain, Fire Suppression   73355.21   
120867     Michael J Moran                              Captain 3     7660.0   
122361  Michael V Rolovich              Captain, Fire Suppression        0.0   
126588    Stephen M Tittel                              Captain 3   68491.89   
132232     Jeanne M Seyler              Captain, Fire Suppression   61039.43   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
1        245131.88  137811.38       NaN  538909.28         538909.28  2011  
2        106088.18    16452.6       NaN  335279.91         335279.91  2011  
11        87082.62   110804.3       NaN  297608.92         297608.92  2011  
17       119397.26   18625.08       NaN  278569.21         278569.21  2011  
22        93200.58   39955.25       NaN  273702.71         273702.71  2011  
...            ...        ...       ...        ...               ...   ...  
116604    18517.18   49917.73  25608.77  141790.12         167398.89  2014  
120867     1496.09  126560.36   2189.51  135716.45         137905.96  2014  
122361         0.0  130320.75       0.0  130320.75         130320.75  2014  
126588         0.0   23721.91  19020.79   92213.80         111234.59  2014  
132232     2285.30    7548.29  20113.76   70873.02          90986.78  2014  

[552 rows x 9 columns]

In [38]:
len(data[data['JobTitle'].str.contains('CAPTAIN',case=False)])

552

In [39]:
data[data['JobTitle'].str.contains('CAPTAIN',case=False)].count()

EmployeeName        552
JobTitle            552
BasePay             551
OvertimePay         552
OtherPay            552
Benefits            411
TotalPay            552
TotalPayBenefits    552
Year                552
dtype: int64

## Display all the names from the fire department:

In [40]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [41]:
data['JobTitle']

0         GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
1                        CAPTAIN III (POLICE DEPARTMENT)
2                        CAPTAIN III (POLICE DEPARTMENT)
3                   WIRE ROPE CABLE MAINTENANCE MECHANIC
4           DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)
                               ...                      
148649                                         Custodian
148650                                      Not provided
148651                                      Not provided
148652                                      Not provided
148653                        Counselor, Log Cabin Ranch
Name: JobTitle, Length: 148654, dtype: object

In [42]:
data['JobTitle'].str.contains('FIRE',case=False)

0         False
1         False
2         False
3         False
4          True
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [43]:
data[data['JobTitle'].str.contains('FIRE',case=False)]

EmployeeName                                          JobTitle  \
4          PATRICK GARDNER      DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
6                ALSON LEE                BATTALION CHIEF, (FIRE DEPARTMENT)   
8           MICHAEL MORRIS                BATTALION CHIEF, (FIRE DEPARTMENT)   
9       JOANNE HAYES-WHITE            CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
10           ARTHUR KENNEY  ASSISTANT CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
...                    ...                                               ...   
145956    Kenneth C Farris                                       Firefighter   
147556       Edward A Dunn                                       Firefighter   
148021      Kari A Johnson                                       Firefighter   
148209        Sheryl K Lee                                       Firefighter   
148554     Lawrence F Gatt                             Fire Alarm Dispatcher   

          BasePay OvertimePay   OtherPay Benefits   TotalPay  \
4        134401.6      9737.0  182234.59      NaN  326373.19   
6        92492.01     89062.9  134426.14      NaN  315981.05   
8       176932.64    86362.68   40132.23      NaN  303427.55   
9        285262.0         0.0   17115.73      NaN  302377.73   
10      194999.39    71344.88    33149.9      NaN  299494.17   
...           ...         ...        ...      ...        ...   
145956       0.00        0.00       0.00  4645.56       0.00   
147556    1063.24        0.00     132.90   385.66    1196.14   
148021     688.71        0.00       0.00   143.39     688.71   
148209     459.14        0.00       0.00    95.59     459.14   
148554      73.33        0.00       0.00     0.73      73.33   

        TotalPayBenefits  Year  
4              326373.19  2011  
6              315981.05  2011  
8              303427.55  2011  
9              302377.73  2011  
10             299494.17  2011  
...                  ...   ...  
145956           4645.56  2014  
147556           1581.80  2014  
148021            832.10  2014  
148209            554.73  2014  
148554             74.06  2014  

[5879 rows x 9 columns]

In [44]:
data[data['JobTitle'].str.contains('FIRE',case=False)]['EmployeeName']

4            PATRICK GARDNER
6                  ALSON LEE
8             MICHAEL MORRIS
9         JOANNE HAYES-WHITE
10             ARTHUR KENNEY
                 ...        
145956      Kenneth C Farris
147556         Edward A Dunn
148021        Kari A Johnson
148209          Sheryl K Lee
148554       Lawrence F Gatt
Name: EmployeeName, Length: 5879, dtype: object

## Find minimum maximum and average basepay column:

In [45]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [46]:
data['BasePay']

0            167411.18
1            155966.02
2            212739.13
3              77916.0
4             134401.6
              ...     
148649            0.00
148650    Not Provided
148651    Not Provided
148652    Not Provided
148653            0.00
Name: BasePay, Length: 148654, dtype: object

In [47]:
data['BasePay'].describe()

count     148049.0
unique    109900.0
top            0.0
freq         875.0
Name: BasePay, dtype: float64

In [48]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   EmployeeName      148654 non-null  object 
 1   JobTitle          148654 non-null  object 
 2   BasePay           148049 non-null  object 
 3   OvertimePay       148654 non-null  object 
 4   OtherPay          148654 non-null  object 
 5   Benefits          112495 non-null  object 
 6   TotalPay          148654 non-null  float64
 7   TotalPayBenefits  148654 non-null  float64
 8   Year              148654 non-null  int64  
dtypes: float64(2), int64(1), object(6)
memory usage: 10.2+ MB


### As we can see that the BasePay column is not showing min max etx because it is a object type column.

### Trying to change the not provided option to NaN and hen try to use the functions:

In [49]:
import numpy as np

In [50]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [51]:
data['BasePay']

0            167411.18
1            155966.02
2            212739.13
3              77916.0
4             134401.6
              ...     
148649            0.00
148650    Not Provided
148651    Not Provided
148652    Not Provided
148653            0.00
Name: BasePay, Length: 148654, dtype: object

In [52]:
data['BasePay'].replace('Not Provided',np.nan)

0         167411.18
1         155966.02
2         212739.13
3           77916.0
4          134401.6
            ...    
148649         0.00
148650          NaN
148651          NaN
148652          NaN
148653         0.00
Name: BasePay, Length: 148654, dtype: object

In [53]:
data['BasePay']=data['BasePay'].replace('Not Provided',np.nan)

In [54]:
data['BasePay']

0         167411.18
1         155966.02
2         212739.13
3           77916.0
4          134401.6
            ...    
148649         0.00
148650          NaN
148651          NaN
148652          NaN
148653         0.00
Name: BasePay, Length: 148654, dtype: object

In [55]:
data['BasePay'].describe()

count     148045.0
unique    109899.0
top            0.0
freq         875.0
Name: BasePay, dtype: float64

### Still it is not working because series is in format of object type hence need to be parsed in int or float.

In [56]:
#Typecasting as float:
data['BasePay'].astype(float)

0         167411.18
1         155966.02
2         212739.13
3          77916.00
4         134401.60
            ...    
148649         0.00
148650          NaN
148651          NaN
148652          NaN
148653         0.00
Name: BasePay, Length: 148654, dtype: float64

In [57]:
data['BasePay']=data['BasePay'].astype(float)

In [58]:
data['BasePay']

0         167411.18
1         155966.02
2         212739.13
3          77916.00
4         134401.60
            ...    
148649         0.00
148650          NaN
148651          NaN
148652          NaN
148653         0.00
Name: BasePay, Length: 148654, dtype: float64

In [59]:
data['BasePay'].describe()

count    148045.000000
mean      66325.448840
std       42764.635495
min        -166.010000
25%       33588.200000
50%       65007.450000
75%       94691.050000
max      319275.010000
Name: BasePay, dtype: float64

In [60]:
data['BasePay'].max()

319275.01

In [61]:
data['BasePay'].min()

-166.01

In [62]:
data['BasePay'].mean()

66325.4488404877

### Now working fine.

## Find minimum maximum and average totalpay column:

In [63]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [64]:
data['TotalPay']

0         567595.43
1         538909.28
2         335279.91
3         332343.61
4         326373.19
            ...    
148649         0.00
148650         0.00
148651         0.00
148652         0.00
148653      -618.13
Name: TotalPay, Length: 148654, dtype: float64

In [65]:
data['TotalPay'].describe()

count    148654.000000
mean      74768.321972
std       50517.005274
min        -618.130000
25%       36168.995000
50%       71426.610000
75%      105839.135000
max      567595.430000
Name: TotalPay, dtype: float64

In [66]:
data['TotalPay'].min()

-618.13

In [67]:
data['TotalPay'].max()

567595.43

In [68]:
data['TotalPay'].mean()

74768.32197169267

### This can be easily calculated as the series type is a float 

## Replacing the not provided from name section to Nan:

In [69]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [70]:
data['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650         Not provided
148651         Not provided
148652         Not provided
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [71]:
data['EmployeeName'].replace('Not provided',np.nan)

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [72]:
# nan= null not a string.
# To store it we need to equate it to data.
data['EmployeeName']=data['EmployeeName'].replace('Not provided',np.nan)

In [73]:
data['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

## Drop the rows having 5 missing values:

In [74]:
data.isnull().sum()
# This gives null data in terms of columns.

EmployeeName            4
JobTitle                0
BasePay               609
OvertimePay             0
OtherPay                0
Benefits            36159
TotalPay                0
TotalPayBenefits        0
Year                    0
dtype: int64

In [75]:
# In terms of rows:
data.isnull().sum(axis=1)

0         1
1         1
2         1
3         1
4         1
         ..
148649    0
148650    2
148651    2
148652    2
148653    0
Length: 148654, dtype: int64

In [76]:
data.isnull().sum(axis=1)==5

0         False
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Length: 148654, dtype: bool

In [77]:
data[data.isnull().sum(axis=1)==5]
# No rows to show

Empty DataFrame
Columns: [EmployeeName, JobTitle, BasePay, OvertimePay, OtherPay, Benefits, TotalPay, TotalPayBenefits, Year]
Index: []

In [78]:
data.drop(data[data.isnull().sum(axis=1)==5].index,axis=0,inplace=True)
#axis=1 for sum for rows
#axis=0 for rows while dropping
#inplace=True to save the changes in the dataframe.

In [79]:
data[data.isnull().sum(axis=1)==5]

Empty DataFrame
Columns: [EmployeeName, JobTitle, BasePay, OvertimePay, OtherPay, Benefits, TotalPay, TotalPayBenefits, Year]
Index: []

## Find the job title of Albert Pardini:

In [80]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [81]:
data['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [82]:
data['EmployeeName']=='ALBERT PARDINI'

0         False
1         False
2          True
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: EmployeeName, Length: 148654, dtype: bool

In [83]:
data[data['EmployeeName']=='ALBERT PARDINI']

EmployeeName                         JobTitle    BasePay OvertimePay  \
2  ALBERT PARDINI  CAPTAIN III (POLICE DEPARTMENT)  212739.13   106088.18   

  OtherPay Benefits   TotalPay  TotalPayBenefits  Year  
2  16452.6      NaN  335279.91         335279.91  2011

In [84]:
data[data['EmployeeName']=='ALBERT PARDINI']['JobTitle']

2    CAPTAIN III (POLICE DEPARTMENT)
Name: JobTitle, dtype: object

In [85]:
# For case insensitive:
data['EmployeeName'].str.contains('ALBERT PARDINI',case=False)

0         False
1         False
2          True
3         False
4         False
          ...  
148649    False
148650      NaN
148651      NaN
148652      NaN
148653    False
Name: EmployeeName, Length: 148654, dtype: object

### data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False)]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [103], line 1
----> 1 data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False)]

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\frame.py:3794, in DataFrame.__getitem__(self, key)
   3791     return self.where(key)
   3793 # Do we have a (boolean) 1d indexer?
-> 3794 if com.is_bool_indexer(key):
   3795     return self._getitem_bool_array(key)
   3797 # We are left with two options: a single key, and a collection of keys,
   3798 # We interpret tuples as collections only for non-MultiIndex

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\common.py:135, in is_bool_indexer(key)
    131     na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
    132     if lib.infer_dtype(key_array) == "boolean" and isna(key_array).any():
    133         # Don't raise on e.g. ["A", "B", np.nan], see
    134         #  test_loc_getitem_list_of_labels_categoricalindex_with_na
--> 135         raise ValueError(na_msg)
    136     return False
    137 return True

### ValueError: Cannot mask with non-boolean array containing NA / NaN values

### This happened because the series of boolean contains Nan values hence we need to use na=False.

In [86]:
data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False,na=False)]

EmployeeName                         JobTitle    BasePay OvertimePay  \
2      ALBERT PARDINI  CAPTAIN III (POLICE DEPARTMENT)  212739.13   106088.18   
36519  Albert Pardini                        Captain 3  196265.04         0.0   

       OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
2       16452.6       NaN  335279.91         335279.91  2011  
36519  11404.04  46326.21  207669.08         253995.29  2012

In [87]:
data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False,na=False)]['JobTitle']

2        CAPTAIN III (POLICE DEPARTMENT)
36519                          Captain 3
Name: JobTitle, dtype: object

## How much does Albert Pardini earn including benifits:

In [88]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [89]:
data['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [90]:
data['EmployeeName'].str.contains('ALBERT PARDINI',case=False,na=False)

0         False
1         False
2          True
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: EmployeeName, Length: 148654, dtype: bool

In [91]:
data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False,na=False)]

EmployeeName                         JobTitle    BasePay OvertimePay  \
2      ALBERT PARDINI  CAPTAIN III (POLICE DEPARTMENT)  212739.13   106088.18   
36519  Albert Pardini                        Captain 3  196265.04         0.0   

       OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
2       16452.6       NaN  335279.91         335279.91  2011  
36519  11404.04  46326.21  207669.08         253995.29  2012

In [92]:
data[data['EmployeeName'].str.contains('ALBERT PARDINI',case=False,na=False)]['TotalPayBenefits']

2        335279.91
36519    253995.29
Name: TotalPayBenefits, dtype: float64

## Display the name of the person having highest and lowest base pay:

In [93]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [94]:
data['BasePay']

0         167411.18
1         155966.02
2         212739.13
3          77916.00
4         134401.60
            ...    
148649         0.00
148650          NaN
148651          NaN
148652          NaN
148653         0.00
Name: BasePay, Length: 148654, dtype: float64

In [95]:
data['BasePay'].max()

319275.01

In [96]:
data['BasePay'].max()==data['BasePay']

0         False
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: BasePay, Length: 148654, dtype: bool

In [97]:
data[data['BasePay'].max()==data['BasePay']]

EmployeeName         JobTitle    BasePay OvertimePay  OtherPay  \
72925  Gregory P Suhr  Chief of Police  319275.01         0.0  20007.06   

       Benefits   TotalPay  TotalPayBenefits  Year  
72925  86533.21  339282.07         425815.28  2013

In [98]:
data[data['BasePay'].max()==data['BasePay']]['EmployeeName']

72925    Gregory P Suhr
Name: EmployeeName, dtype: object

## Find the Average BasePay for all the Employees per year:

In [99]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [100]:
data.groupby('Year').mean()['BasePay']
# C:\Users\versha pande\AppData\Local\Temp\ipykernel_47200\3282093210.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
#   data.groupby('Year').mean()['BasePay']
# This happens because there are NaN values here.

C:\Users\versha pande\AppData\Local\Temp\ipykernel_47296\3798292640.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  data.groupby('Year').mean()['BasePay']


Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [101]:
data.groupby('Year').mean(numeric_only=True)['BasePay']
# Without any warning.

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

## Find the Average BasePay for all the Employees per job title:

In [102]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [103]:
data.groupby('JobTitle').mean(numeric_only=True)['BasePay']

JobTitle
ACCOUNT CLERK                                     43300.806506
ACCOUNTANT                                        46643.172000
ACCOUNTANT INTERN                                 28732.663958
ACPO,JuvP, Juv Prob (SFERS)                       62290.780000
ACUPUNCTURIST                                     66374.400000
                                                      ...     
X-RAY LABORATORY AIDE                             47664.773077
X-Ray Laboratory Aide                             46086.387100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.910000
Youth Comm Advisor                                39077.957500
ZOO CURATOR                                       43148.000000
Name: BasePay, Length: 2159, dtype: float64

## Find the Average BasePay for all the Employees having job title Accountant:

In [104]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [105]:
data['JobTitle'].str.contains('Accountant',case=False)

0         False
1         False
2         False
3         False
4         False
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [106]:
data[data['JobTitle'].str.contains('Accountant',case=False)]

EmployeeName                   JobTitle    BasePay OvertimePay  \
7444       VICTORIA SANTOS  SENIOR SYSTEMS ACCOUNTANT  107382.02         0.0   
7686           CAROLYN LEW  SENIOR SYSTEMS ACCOUNTANT  109158.60         0.0   
7687              KAN HTUN  SENIOR SYSTEMS ACCOUNTANT  109158.06         0.0   
8158           EDUARDO LIM  SENIOR SYSTEMS ACCOUNTANT  106492.07         0.0   
8162    FLORDELIZA LEVISTE  SENIOR SYSTEMS ACCOUNTANT  106492.04         0.0   
...                    ...                        ...        ...         ...   
143492     Theodore N Wong              Accountant IV   12019.62        0.00   
143766   Victoria L Santos              Accountant IV   10966.80        0.00   
144811   Violeta M Mosuela              Accountant II    7836.41        0.00   
146288           Dee P Ngo             Accountant III    3087.00        0.00   
148174         Evan J Chan             Accountant III       0.00        0.00   

       OtherPay Benefits   TotalPay  TotalPayBenefits  Year  
7444     3404.0      NaN  110786.02         110786.02  2011  
7686        0.0      NaN  109158.60         109158.60  2011  
7687        0.0      NaN  109158.06         109158.06  2011  
8158        0.0      NaN  106492.07         106492.07  2011  
8162        0.0      NaN  106492.04         106492.04  2011  
...         ...      ...        ...               ...   ...  
143492     0.00     0.00   12019.62          12019.62  2014  
143766     0.00   109.66   10966.80          11076.46  2014  
144811     0.00    78.36    7836.41           7914.77  2014  
146288     0.00   723.28    3087.00           3810.28  2014  
148174   510.00    97.36     510.00            607.36  2014  

[1010 rows x 9 columns]

In [107]:
data[data['JobTitle'].str.contains('Accountant',case=False)]['BasePay'].mean()

72245.57241071429

## Display top 5 common jobs:

In [108]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [109]:
data['JobTitle']

0         GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
1                        CAPTAIN III (POLICE DEPARTMENT)
2                        CAPTAIN III (POLICE DEPARTMENT)
3                   WIRE ROPE CABLE MAINTENANCE MECHANIC
4           DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)
                               ...                      
148649                                         Custodian
148650                                      Not provided
148651                                      Not provided
148652                                      Not provided
148653                        Counselor, Log Cabin Ranch
Name: JobTitle, Length: 148654, dtype: object

In [110]:
data['JobTitle'].value_counts()

Transit Operator                                  7036
Special Nurse                                     4389
Registered Nurse                                  3736
Public Svc Aide-Public Works                      2518
Police Officer 3                                  2421
                                                  ... 
CHIEF HOUSING INSPECTOR                              1
TRAFFIC SIGNAL OPERATOR                              1
COURT COMPUTER FACILITIES COORDINATOR                1
AUTOMOTIVE BODY AND FENDER WORKER SUPERVISOR I       1
VICTIM & WITNESS TECHNICIAN                          1
Name: JobTitle, Length: 2159, dtype: int64

In [111]:
data['JobTitle'].value_counts().head(5)

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64